File: rd-net/RdFramework/ClientId.cs
using System;
using JetBrains.Annotations;
using JetBrains.Rd.Impl;

namespace JetBrains.Rd
{
    /// <summary>
    /// ClientId is a global context class that is used to distinguish the originator of an action in multi-client systems
    /// In such systems, each client has their own ClientId.
    /// 
    /// The context is automatically propagated across async/await calls using AsyncLocal. The application should take care to preserve and propagate the current value across other kinds of asynchronous calls. 
    /// </summary>
    public struct ClientId : IEquatable<ClientId>
    {
        [NotNull] public readonly string Value;
        
        public ClientId([NotNull] string value)
        {
            Value = value;
        }

        public enum AbsenceBehavior
        {
            RETURN_LOCAL,
            THROW
        }

        public static AbsenceBehavior AbsenceBehaviorValue = AbsenceBehavior.RETURN_LOCAL;

        public override string ToString()
        {
            return $"ClientId({Value})";
        }

        public static readonly ClientId LocalId = new ClientId("Host");
        
        public static readonly RdContext<string> Context = new RdContext<string>("ClientId", true, Serializers.ReadString, Serializers.WriteString);

        public static readonly CtxReadDelegate<ClientId> ReadDelegate = (ctx, reader) => new ClientId(reader.ReadString());
        public static readonly CtxWriteDelegate<ClientId> WriteDelegate = (ctx, writer, value) => writer.Write(value.Value);


        #region Cookie

        public struct ClientIdCookie : IDisposable
        {
            private readonly ClientId? myOldClientId;

            public ClientIdCookie(ClientId? newClientId)
            {
                myOldClientId = CurrentOrNull;
                SetClientId(newClientId);
            }

            private static void SetClientId(ClientId? newClientId)
            {
              Context.Value = newClientId?.Value;
            }

            public void Dispose()
            {
                SetClientId(myOldClientId);
            }
        }

        #endregion

        public static ClientIdCookie CreateCookie(ClientId? clientId) => new ClientIdCookie(clientId);

        public static ClientId Current
        {
            get
            {
                switch (AbsenceBehaviorValue)
                {
                    case AbsenceBehavior.RETURN_LOCAL:
                        return CurrentOrNull ?? LocalId;
                    case AbsenceBehavior.THROW:
                        return CurrentOrNull ?? throw new NullReferenceException("ClientId not set");
                    default:
                        throw new ArgumentOutOfRangeException(nameof(AbsenceBehaviorValue));
                }
            }
        }

        [CanBeNull]
        public static ClientId? CurrentOrNull
        {
          get
          {
            var contextValue = Context.Value;
            if (contextValue == null)
              return null;
            return new ClientId(contextValue);
          }
        }

        #region Equality members

        public bool Equals(ClientId other)
        {
            return Value == other.Value;
        }

        public override bool Equals(object obj)
        {
            return obj is ClientId other && Equals(other);
        }

        public override int GetHashCode()
        {
            return Value.GetHashCode();
        }

        public static bool operator ==(ClientId left, ClientId right)
        {
            return left.Equals(right);
        }

        public static bool operator !=(ClientId left, ClientId right)
        {
            return !left.Equals(right);
        }

        #endregion
    }
}

File: rd-kt/rd-framework/src/commonMain/kotlin/com/jetbrains/rd/framework/ClientId.kt
package com.jetbrains.rd.framework

import com.jetbrains.rd.util.Callable
import com.jetbrains.rd.util.Runnable
import com.jetbrains.rd.util.reflection.usingValue
import kotlin.jvm.JvmStatic

/**
 * ClientId is a global context class that is used to distinguish the originator of an action in multi-client systems
 * In such systems, each client has their own ClientId. Current process also can have its own ClientId, with this class providing methods to distinguish local actions from remote ones.
 *
 * It's up to the application to preserve and propagate the current value across background threads and asynchronous activities.
 */
data class ClientId(val value: String) {
    enum class AbsenceBehavior {
        /**
         * Return localId if ClientId is not set
         */
        RETURN_LOCAL,
        /**
         * Throw an exception of ClientId is not set
         */
        THROW
    }

    companion object : ISerializer<ClientId> {
        private val defaultLocalId = ClientId("Host")
        val contextKey = RdContext("ClientId", true, FrameworkMarshallers.String)

        /**
         * Specifies behavior for ClientId.current
         */
        var AbsenceBehaviorValue = AbsenceBehavior.RETURN_LOCAL

        /**
         * The ID considered local to this process. All other IDs (except for null) are considered remote
         */
        @JvmStatic
        var localId = defaultLocalId
            get
            private set

        /**
         * True if and only if the current ClientID is local to this process
         */
        @JvmStatic
        val isCurrentlyUnderLocalId: Boolean
            get() = isLocalId(current)

        /**
         * Gets the current ClientId. Subject to AbsenceBehaviorValue
         */
        @JvmStatic
        val current: ClientId
            get() = when(AbsenceBehaviorValue) {
                AbsenceBehavior.RETURN_LOCAL -> currentOrNull ?: localId
                AbsenceBehavior.THROW -> currentOrNull ?: throw NullPointerException("ClientId not set")
            }

        /**
         * Gets the current ClientId. Can be null if non was set.
         */
        @JvmStatic
        val currentOrNull: ClientId?
            get() = contextKey.value?.let(::ClientId)

        /**
         * Overrides the ID that is considered to be local to this process. Can be only invoked once.
         */
        @JvmStatic
        fun overrideLocalId(newId: ClientId) {
            require(localId == defaultLocalId)
            localId = newId
        }

        /**
         * Returns true if and only if the given ID is considered to be local to this process
         */
        @JvmStatic
        fun isLocalId(clientId: ClientId?): Boolean {
            return clientId == localId || clientId == null
        }

        /**
         * Invokes a runnable under the given ClientId
         */
        @JvmStatic
        fun withClientId(clientId: ClientId?, action: Runnable) = withClientId(clientId) { action.run() }

        /**
         * Computes a value under given ClientId
         */
        @JvmStatic
        fun <T> withClientId(clientId: ClientId?, action: Callable<T>): T = withClientId(clientId) { action.call() }

        /**
         * Computes a value under given ClientId
         */
        @JvmStatic
        inline fun <T> withClientId(clientId: ClientId?, action: () -> T): T = contextKey::value.usingValue(clientId?.value, action)

        override fun read(ctx: SerializationCtx, buffer: AbstractBuffer): ClientId {
            return ClientId(buffer.readString())
        }

        override fun write(ctx: SerializationCtx, buffer: AbstractBuffer, value: ClientId) {
            buffer.writeString(value.value)
        }
    }
}